### PR TITLE
New Storage

### LINKED ISSUE #3: Framing
currently the framing will not work for long log statements.

Blocked by: https://github.com/knurling-rs/defmt/issues/185

### LINKED ISSUE #5: COBS encode from Postcard
Currently the COBS encoding decoding is done "manually", but should be possible to get working as a `postcard` flavor

### LINKED ISSUE #8: Update to crates.io version of defmt
This should be updated to use https://crates.io/crates/defmt

### LINKED ISSUE #10: Periodic persisting of log records fails
Hello @MathiasKoch! Thanks for this crate, I find it useful and this is exactly what I had in mind for my project for quite some time!

I'm currently trying to plug this into my project that uses an STM32 MCU that has a typical FLASH controller that requires the page to be erased to be written to.

From my understanding, this requirement poses a challenge, that it isn't possible to periodically call `drain_storage` to flush records into the FLASH as it will fail on consequent calls due to the page with the header not being erased.

Currently, I can imagine some solutions to this:

1. Keep the header in a separate FLASH page and erase it every time the header needs to be updated. This solution is pretty lame and wasteful of the FLASH resource but could work in the short term, or
2. Do not use the header and rely on the fact that "empty" data in FLASH are always `0xFF`s. Probably, the log records could have some magic number at the beginning (and/or the end) and during the write, we will skim through the log area to find an empty spot that is only `0xFF`s and doesn't have that magic number at the beginning, or
3. Use some more advanced data structure that is more suitable for a FLASH memory, probably similar to those used in filesystems

Hope it makes sense. What are your thoughts on this?

### PR DESCRIPTION
**Pull request by eupn**
Hi @MathiasKoch,

As discussed in #10, we have to avoid the usage of a header due to properties and limitations posed by the internal (and maybe external too) FLASH memories which are the most probable target use case of this crate.

The initial idea implemented here is kinda dumb and simple:

* Have a "magic" word at the very beginning of the storage area as a sign that the area is initialized
  * If there's no magic word, then write the magic word and consider the storage area empty
    * Set both read and write "heads" right after the magic word
  * If the magic word is present, then seek for the first two consecutive "empty" words (i.e. `0xFFFF_FFFF 0xFFFF_FFFF`)
    * If such two words were found set the "write" head at the beginning of the first "empty" word
    * If no such two words found, then consider the storage full. (We need to further think about how to handle wrap-around of logging area)
*  When flushing log frames from the buffer (by calling `drain_storage`), encode each frame with a modified version of COBS that has `0xFF` byte as a sentinel byte instead of standard `0x00`. By doing this we can be sure that we'll be able to find two consecutive "empty" (`0xFF`s) words as a next spot to write logs to should the system be restarted. COBS guarantees that `0xFF` bytes won't appear within a legit log frame so this must be an empty space.
* When reading log frames from the storage:
  * Split frames by `0xFF` byte since this is our modified-COBS sentinel byte
  * Due to the properties of the FLASH, there may be multiple `0xFF` bytes in a row. We'll consider these sequences as COBS frames of the length of 1 and skip these
  *  We'll count a number of consecutive "empty" frames and if this number reaches 7, we'll abort reading as we're now sure that we reached the end of the current log stream (if the wrap-around possibility is introduced, this should be revised)
  * COBS-decode each frame to get the `defmt` frame

For a better review experience, I think it'd better review this as a whole and not commit-by-commit.

I'm marking this as a WIP for now as I'm currently testing this in hardware, but nevertheless, I'm interested to hear your opinion.

Fixes #8 
Fixes #10 
Fixes #3 
Fixes #5

### CODE REVIEW

**Comment by MathiasKoch** [comment]
I think if we change the cobs encoder to start in `aqcuire` and finalize in `release`, then my original concern around multiple writes per log frame shouldn't be an issue on this. 

Another this is that i think we can improve storage a bit by using https://github.com/Dirbaio/rzcobs instead of regular cobs?

**Comment by eupn** [comment]
@MathiasKoch yes, I think given the new information, I should revise the code a bit to have a stateful encoder that encodes for multiple `write`s in between `acquire`/`release` calls.

>Another this is that i think we can improve storage a bit by using https://github.com/Dirbaio/rzcobs instead of regular cobs, to gain zero compression?

I think yes, in this regard it's promising, as we strive to safe as much space as possible. However, I'm not sure if it'll work properly with non-zero sentinel byte (in our case it's `0xFF`). The current way we deal with this byte is by just XOR-ing the COBS-encoded bytes with `0xFF` and I need to check if this will work with rzCOBS. Most likely, yes.

**Comment by MathiasKoch** [comment]
Ahh yeah, i see that neither rcobs nor rzcobs does initially offer a custom sentinel marker. Not sure if this is possible and just not implemented, or if the reverse stuff makes it impossible for some reason? @Dirbaio might have some insights on that?

@eupn Would you mind sharing a few words about your usecase of this crate? What kind of transport do you intend to put on top, if any? Eg in our application we intend to put a pub/sub mqtt layer ontop to allow on-demand retreival of the log history.

**Comment by eupn** [comment]
@MathiasKoch my use case is logging for a wearable device that logs some stuff into internal FLASH memory. Mostly for debugging and post-mortems. So logs and panic messages (if any) should be preserved to then being read out with the debugger probe or via USB. Debugger probe is an easier option for me, as then I can just patch the `probe-run` tool to just connect to the device, read FLASH area with logs and decode the `defmt` frames and display those correctly as `probe-run` already has the machinery to format log strings by using data from the executable.

**Comment by MathiasKoch** [comment]
That sounds cool!

> P.S.: in my opinion, this crate should be agnostic of any transport used and only take care of making sure that defmt frames are persisted. How to further transport them is kinda out of scope.

I completely agree on this, i was meerly curious if had overseen some hidden potential ;)

**Comment by eupn** [comment]
@MathiasKoch 

I've verified that this works in hardware and in the process I've made some fixes and a small [tool](https://github.com/eupn/defmt-print) that receives `probe-rs-cli`'s FLASH area dump as well as firmware ELF and decodes FLASH data into human-readable `defmt` logs. For your MQTT case, I'm imagining that you can request persisted logs from the device and pipe them into that tool (provided the data has the same format as `probe-rs` uses, though it shouldn't be too hard to tweak the data a bit to be accepted by the tool), so that tool may be also useful for you.

I've modified the code to take into account that we can accumulate `write`s in between `acquire` and `release` calls though it's now more prone to panic.

File: src/producer.rs [hunk 1]
```diff
@@ -9,6 +9,8 @@ use cortex_m::{interrupt, register};
 #[defmt::global_logger]
 pub struct Logger;
 
+defmt::timestamp!("");
+
 impl defmt::Write for Logger {
     // fn start_of_frame(&mut self, len: usize) {
```
Comment (MathiasKoch) [inline src/producer.rs:15]: Could you remove these commented out functions? 
They were just a reminder of the multi-write issue, which you have now fixed :+1:

File: src/lib.rs [hunk 2]
```diff
@@ -363,23 +354,42 @@ where
     /// storage.
     ///
     /// Returns the number of bytes pushed to `buf`
-    pub fn retreive_frames(&mut self, storage: &mut S, buf: &mut [u8]) -> Result<usize, Error> {
-        let read_len = self.helper.read_slice(storage, buf)?;
+    pub fn retrieve_frames(&mut self, storage: &mut S, buf: &mut [u8]) -> Result<usize, Error> {
```
Comment (MathiasKoch) [inline src/lib.rs:357]: Is there a reason for the `retrieve_frames_helper` vs `retrieve_frames` ? Couldn't they just be combined into `retrieve_frames`?

**Review by MathiasKoch — CHANGES_REQUESTED** [review]
Minor nitpicks on just reading through the code.

One thought is, perhaps we should remove the `defmt-rtt` part entirely? The idea behind it was just an easy way to switch between storing logs in storage, and directing them to rtt. But it think perhaps we should leave that feature gating to a user level crate?

File: src/producer.rs (hunk 1 above)
Comment (eupn) [reply src/producer.rs:15]: Sure!

File: src/lib.rs (hunk 2 above)
Comment (eupn) [reply src/lib.rs:357]: It was easier for [defmt-print](https://github.com/eupn/defmt-print/blob/4fcaf42a41b14041c357bf22fb1109753ac963a0/src/main.rs#L70) to reuse the code from this crate if we extract frames reading logic into a static helper function like this. This way it won't require `defmt-print` to instantiate `LogManager` instance (as well as `bbqueue` buffer) only for a purpose of reading log frames.

Comment (MathiasKoch) [reply src/lib.rs:357]: Ahh, okay :+1: Fair enough.

As long as there was reasoning behind it :)

File: src/producer.rs (hunk 1 above)
Comment (eupn) [reply src/producer.rs]: @MathiasKoch done

**Comment by eupn** [comment]
@MathiasKoch 

>One thought is, perhaps we should remove the defmt-rtt part entirely?

Yes, or can we have both RTT and persisting? It could be useful to run the firmware with `probe-run`, see some logs there that it's running OK, and then remove the probe or deploy the device in the field while still be able to read logs, but later and from storage. Both achieved by the same firmware.

**Review by MathiasKoch — APPROVED** [review]

**Pull request merged**